### ISSUE TITLE
formats loop chewing RAM for non SvPVs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13277]
Migrated from [rt.perl.org#119847](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=119847) (status was 'resolved')

Searchable as RT119847$

**Comment by p5pRT** [comment]
### From @nwc10

formats loop forever chewing RAM for non SVPVs\. This isn't a recent problem\,
but a couple of \(relatively\) recent commits have exacerbated it\.

Most recently\, this code\, adapted from t/op/write\.t\, will now eat RAM&#8203;:

\#\!\./perl

use strict;     \# Amazed that this hackery can be made strict \.\.\.

my $v = 0\.632120558828558;
\# use Devel&#8203;::Peek;
\# Dump $v;

format STDOUT =
ok ^\<\<\<\<\<\<\<\<\< ~~
$v
\.

write STDOUT;

\# Dump $v;



This used to work until this commit&#8203;:

commit b127e37e51c21b0a36755dcd19811be931a03d83
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Mon May 13 07&#8203;:35&#8203;:35 2013 \-0600

&nbsp;    PATCH&#8203;: \[perl \#108378\] \[perl \#115800\]

&nbsp;    This patch solves two tickets\.  Both are a result of the stringification
&nbsp;    of a floating number being sticky\, so that the character representing
&nbsp;    the decimal point may be from an old locale\.  The patch solves this by
&nbsp;    not retaining the stringification of NVs\.


There is nothing wrong with that commit \- it just exposes the problem\.
Using a tied variable or overloading to enforce an NV every time has failed
since this commit&#8203;:

commit dc35ab6e9838269debf9973a573bbd31031f3f31
Author&#8203;: Rafael Garcia\-Suarez \<rgs@&#8203;consttype\.org>
Date&#8203;:   Thu Oct 22 23&#8203;:04&#8203;:14 2009 \+0200

&nbsp;    Remove the venerable fatal error "Runaway format"
&nbsp;    
&nbsp;    This solves bug "\[perl \#69927\] wrong runaway error for write"\,
&nbsp;    which is a link to http&#8203;://bugs\.debian\.org/cgi\-bin/bugreport\.cgi?bug=77707


Previously it was trapped by the "Runaway format" case\, but again\, I don't
think that that commit is at fault either\. So these two&#8203;:

\#\!\./miniperl

use strict;     \# Amazed that this hackery can be made strict \.\.\.

\{
&nbsp;    package foo;
&nbsp;    use overload "" => sub \{ "Hello world" \};
\}

my $v = bless \[\]\, "foo";
\#use Devel&#8203;::Peek;
\#Dump $v;

format STDOUT =
ok ^\<\<\<\<\<\<\<\<\< ~~
$v
\.

eval \{write STDOUT\}
&nbsp;    or warn $@&#8203;;

\#Dump $v;
\_\_END\_\_


\#\!\./perl

use strict;

\{   
&nbsp;    package Number;
&nbsp;    my $zero = $$ \- $$;

&nbsp;    sub TIESCALAR \{
&nbsp;        my $class = shift;
&nbsp;        my $value = shift;
&nbsp;        return bless \\$value\, $class;
&nbsp;    \}

&nbsp;    \# The return value should always be SvNOK\(\) only&#8203;:
&nbsp;    sub FETCH \{
&nbsp;        my $self = shift;
&nbsp;        return $$self \- 0\.5 \+ $zero \+ 0\.5;
&nbsp;    \}

&nbsp;   sub STORE \{
&nbsp;       my $self = shift;
&nbsp;       $$self = shift;
&nbsp;    \}
\}

tie my $v27c\, 'Number'\, 1/256;

format STDOUT =
'^\<\<\<\<\<\<\<\<\<'~~
$v27c
\.

eval \{write STDOUT\}
&nbsp;    or warn $@&#8203;;
\_\_END\_\_

used to croak with the "Runaway format" error until that commit abolished it\.


In all cases\, I believe that the cause is this part of pp\_formline\,
specifically the call to sv\_chop\(\) and bad assumptions about it&#8203;:

&nbsp;    case FF\_CHOP&#8203;:
&nbsp;        \{
&nbsp;        const char \*s = chophere;
&nbsp;        if \(chopspace\) \{
&nbsp;            while \(isSPACE\(\*s\)\)
&nbsp;            s\+\+;
&nbsp;        \}
&nbsp;        sv\_chop\(sv\,s\);
&nbsp;        SvSETMAGIC\(sv\);
&nbsp;        break;
&nbsp;        \}



By calling sv\_chop\(\) the code \*intends\* to remove from the scalar the part
that has already been output\, so that the next time that pp\_formline is
called\, output will continue from the right place\, and when the scalar has
been exhausted\, output stops\.

The problem is that sv\_chop\(\) starts like this&#8203;:

&nbsp;    if \(\!ptr || \!SvPOKp\(sv\)\)
&nbsp;    return;


returning immediately \*without changing anything\* if the SV is not already
a string\. This doesn't work\, because pp\_formline doesn't force convert the
scalar to a string\, so any \(sticky\) non\-string SV type will never be updated\.
example problem code&#8203;:

&nbsp;    case FF\_LINEGLOB&#8203;:
&nbsp;        \{
&nbsp;        const bool oneline = fpc\[\-1\] == FF\_LINESNGL;
&nbsp;        const char \*s = item = SvPV\_const\(sv\, len\);
&nbsp;        const char \*const send = s \+ len;

&nbsp;        item\_is\_utf8 = DO\_UTF8\(sv\);
&nbsp;        if \(\!len\)
&nbsp;            break;
&nbsp;        trans = 0;
&nbsp;        gotsome = TRUE;
&nbsp;        chophere = s \+ len;
&nbsp;        source = \(U8 \*\) s;
&nbsp;        to\_copy = len;
&nbsp;        while \(s \< send\) \{
&nbsp;            if \(\*s\+\+ == '\\n'\) \{
&nbsp;            if \(oneline\) \{
&nbsp;                to\_copy = s \- SvPVX\_const\(sv\) \- 1;
&nbsp;                chophere = s;
&nbsp;                break;



I think that the right solution would be to replace the simple call to
sv\_chop\(sv\, s\) \[shown above\] with a condition check on SvPOKp\(sv\)\, and if
the check is not true\, call sv\_setpvn\(\) instead\.


There is a non\-zero possibility that existing user code is at risk due to
this\, if upgrading from pre\-5\.12 to 5\.12\, because it will no longer croak\(\)
for some problem cases\. I can't think of a case where existing code that
worked cleanly stops working\, and I think that all upgraded code would be
spotted with sufficient \(probably any\) testing\. Hence I'm wondering if the
risk of subverting this bug is theoretical\, and hence we should treat this
the same way as we handled the bug fixed by&#8203;:

commit 35c6393c7abe46176380cd5f437568a98c8f4995
Author&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
Date&#8203;:   Sun Jul 13 21&#8203;:37&#8203;:21 2008 \+0000

&nbsp;    Make format items @&#8203;\* and ^\* work with references \(safely\)\. Note no\-one
&nbsp;    said anything about sanely\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;34140


\(ie we just fixed it\)


Given that I was thinking that the solution is Perl\_sv\_setpvn\(\)\, I was
wondering what happens if the SV is not the same UTF8\-ness each time\, which
turned out to produce interesting results\. I think that it might be fixed
\(or at least partly fixed\) when the sv\_chop\(\) problem is addressed\. However\,
I'll report it as a separate bug\, because it seems to have fewer untoward
side effects\. \(The combination of the UTF\-8 toggle and an \*overloaded\*
string is a RAM\-chewer\, so I'll attach its tests to this report\)

Attached test cases are intended to be merged into t/op/write\.t once this
is fixed\. I don't have time to investigate this for the foreseeable future\,
so it would be useful if someone else could tackle it \(and say if they do\)\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

\#\!\./perl

BEGIN \{
&nbsp;    chdir 't' if \-d 't';
&nbsp;    @&#8203;INC = '\.\./lib';
&nbsp;    require '\./test\.pl';
\}

use strict;
require Tie&#8203;::Scalar;

\# read in a file
sub cat\_utf8 \{
&nbsp;    my $file = shift;
&nbsp;    local $/;
&nbsp;    local $&#8203;::Level = $&#8203;::Level \+ 1;
&nbsp;    open my $fh\, '\<'\, $file or die "can't open '$file'&#8203;: $\!";
&nbsp;    binmode $fh\, '&#8203;:utf8';
&nbsp;    my $data = \<$fh> // die "Can't read from '$file'&#8203;: $\!";
&nbsp;    close $fh or die "error closing '$file'&#8203;: $\!";
&nbsp;    $data;
\}

sub like\_format\_utf8 \{
&nbsp;    my \($glob\, $want\, $desc\) = @&#8203;\_;
&nbsp;    local $&#8203;::Level = $&#8203;::Level \+ 1;
&nbsp;    my $file = 'Op\_write\.tmp';
&nbsp;    open $glob\, '>&#8203;:utf8'\, $file or die "Can't create '$file'&#8203;: $\!";
&nbsp;    write $glob;
&nbsp;    close $glob or die "Could not close '$file'&#8203;: $\!";
&nbsp;    like\(cat\_utf8\($file\)\, $want\, $desc\);
\}

\{
&nbsp;    package Ref;
&nbsp;    use overload "" => sub \{
&nbsp;    return $\{$\_\[0\]\};
&nbsp;    \};

&nbsp;    sub new \{
&nbsp;       my $class = shift;
&nbsp;       my $value = shift;
&nbsp;       return bless \\$value\, $class;
&nbsp;   \}
\}

plan\(6\);
watchdog\(5\);

my $v28a = \{\};

format OUT28a =
'^\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<'~~
$v28a
\.

like\_format\_utf8\(\\\*OUT28a\,
&nbsp;                 qr/''\\n/\); \# FIXME with correct output once known

my $v28b = \{\};

format OUT28b =
'^\<\<\<'~~
$v28b
\.

like\_format\_utf8\(\\\*OUT28b\,
&nbsp;                 qr/''\\n/\); \# FIXME with correct output once known


tie my $v28c\, 'Tie&#8203;::StdScalar';
$v28c = \{\};

format OUT28c =
'^\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<'~~
$v28c
\.

like\_format\_utf8\(\\\*OUT28c\,
&nbsp;                 qr/''\\n/\); \# FIXME with correct output once known

tie my $v28d\, 'Tie&#8203;::StdScalar';
$v28d = \{\};

format OUT28d =
'^\<\<\<'~~
$v28d
\.

like\_format\_utf8\(\\\*OUT28d\,
&nbsp;                 qr/''\\n/\); \# FIXME with correct output once known

my $v28e = Ref\->new\(\{\}\);

format OUT28e =
'^\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<'~~
$v28e
\.

like\_format\_utf8\(\\\*OUT28e\,
&nbsp;                 qr/''\\n/\); \# FIXME with correct output once known

my $v28f = Ref\->new\(\{\}\);

format OUT28f =
'^\<\<\<'~~
$v28f
\.

like\_format\_utf8\(\\\*OUT28f\,
&nbsp;                 qr/''\\n/\); \# FIXME with correct output once known

**Comment by p5pRT** [comment]
### From @nwc10

\#\!\./perl

BEGIN \{
&nbsp;    chdir 't' if \-d 't';
&nbsp;    @&#8203;INC = '\.\./lib';
&nbsp;    require '\./test\.pl';
\}

use strict;

\# read in a file
sub cat\_utf8 \{
&nbsp;    my $file = shift;
&nbsp;    local $/;
&nbsp;    open my $fh\, '\<'\, $file or die "can't open '$file'&#8203;: $\!";
&nbsp;    binmode $fh\, '&#8203;:utf8';
&nbsp;    my $data = \<$fh> // die "Can't read from '$file'&#8203;: $\!";
&nbsp;    close $fh or die "error closing '$file'&#8203;: $\!";
&nbsp;    $data;
\}

sub is\_format\_utf8 \{
&nbsp;    my \($glob\, $want\, $desc\) = @&#8203;\_;
&nbsp;    local $&#8203;::Level = $&#8203;::Level \+ 1;
&nbsp;    my $file = 'Op\_write\.tmp';
&nbsp;    open $glob\, '>&#8203;:utf8'\, $file or die "Can't create '$file'&#8203;: $\!";
&nbsp;    write $glob;
&nbsp;    close $glob or die "Could not close '$file'&#8203;: $\!";
&nbsp;    is\(cat\_utf8\($file\)\, $want\, $desc\);
\}

plan\(6\);
watchdog\(5\);

\{   
&nbsp;    my $zero = $$ \- $$;

&nbsp;    package Number;

&nbsp;    sub TIESCALAR \{
&nbsp;        my $class = shift;
&nbsp;        my $value = shift;
&nbsp;        return bless \\$value\, $class;
&nbsp;    \}

&nbsp;    \# The return value should always be SvNOK\(\) only&#8203;:
&nbsp;    sub FETCH \{
&nbsp;        my $self = shift;
&nbsp;        return $$self \- 0\.5 \+ $zero \+ 0\.5;
&nbsp;    \}

&nbsp;   sub STORE \{
&nbsp;       my $self = shift;
&nbsp;       $$self = shift;
&nbsp;    \}

&nbsp;   package ONumber;

&nbsp;   use overload "" => sub \{
&nbsp;        my $self = shift;
&nbsp;        return $$self \- 0\.5 \+ $zero \+ 0\.5;
&nbsp;    \};
&nbsp;       
&nbsp;    sub new \{
&nbsp;       my $class = shift;
&nbsp;       my $value = shift;
&nbsp;       return bless \\$value\, $class;
&nbsp;   \}
\}

my $v27a = 1/256;

format OUT27a =
'^\<\<\<\<\<\<\<\<\<'~~
$v27a
\.

is\_format\_utf8\(\\\*OUT27a\,
&nbsp;               "'0\.00390625'\\n"\);

my $v27b = 1/256;

format OUT27b =
'^\<'~~
$v27b
\.

is\_format\_utf8\(\\\*OUT27b\,
&nbsp;               "'0\.'\\n'00'\\n'39'\\n'06'\\n'25'\\n"\);

tie my $v27c\, 'Number'\, 1/256;

format OUT27c =
'^\<\<\<\<\<\<\<\<\<'~~
$v27c
\.

is\_format\_utf8\(\\\*OUT27c\,
&nbsp;               "'0\.00390625'\\n"\);

my $v27d = 1/256;

format OUT27d =
'^\<'~~
$v27d
\.

is\_format\_utf8\(\\\*OUT27d\,
&nbsp;               "'0\.'\\n'00'\\n'39'\\n'06'\\n'25'\\n"\);

my $v27e = ONumber\->new\(1/256\);

format OUT27e =
'^\<\<\<\<\<\<\<\<\<'~~
$v27e
\.

is\_format\_utf8\(\\\*OUT27e\,
&nbsp;               "'0\.00390625'\\n"\);

my $v27f = ONumber\->new\(1/256\);

format OUT27f =
'^\<'~~
$v27f
\.

is\_format\_utf8\(\\\*OUT27f\,
&nbsp;               "'0\.'\\n'00'\\n'39'\\n'06'\\n'25'\\n"\);

**Comment by p5pRT** [comment]
### From @nwc10

\#\!\./perl

BEGIN \{
&nbsp;    chdir 't' if \-d 't';
&nbsp;    @&#8203;INC = '\.\./lib';
&nbsp;    require '\./test\.pl';
\}

use strict;

\# read in a file
sub cat\_utf8 \{
&nbsp;    my $file = shift;
&nbsp;    local $/;
&nbsp;    open my $fh\, '\<'\, $file or die "can't open '$file'&#8203;: $\!";
&nbsp;    binmode $fh\, '&#8203;:utf8';
&nbsp;    my $data = \<$fh> // die "Can't read from '$file'&#8203;: $\!";
&nbsp;    close $fh or die "error closing '$file'&#8203;: $\!";
&nbsp;    $data;
\}

sub is\_format\_utf8 \{
&nbsp;    my \($glob\, $want\, $desc\) = @&#8203;\_;
&nbsp;    local $&#8203;::Level = $&#8203;::Level \+ 1;
&nbsp;    my $file = 'Op\_write\.tmp';
&nbsp;    open $glob\, '>&#8203;:utf8'\, $file or die "Can't create '$file'&#8203;: $\!";
&nbsp;    write $glob;
&nbsp;    close $glob or die "Could not close '$file'&#8203;: $\!";
&nbsp;    is\(cat\_utf8\($file\)\, $want\, $desc\);
\}

plan\(8\);
watchdog\(5\);

\{
&nbsp;    package OS&#8203;::UTF8Toggle;
&nbsp;    use overload "" => sub \{
&nbsp;        my $self = shift;
&nbsp;        $self\->\[1\] = \! $self\->\[1\];
&nbsp;        if \($self\->\[1\]\) \{
&nbsp;            utf8&#8203;::downgrade\($self\->\[0\]\);
&nbsp;        \} else \{
&nbsp;            utf8&#8203;::upgrade\($self\->\[0\]\);
&nbsp;        \}
&nbsp;        $self\->\[0\];
&nbsp;    \};

&nbsp;    sub new \{
&nbsp;        my \($class\, $value\, $state\) = @&#8203;\_;
&nbsp;        bless \[$value\, $state\]\, $class;
&nbsp;    \}
\}


my $u26a = OS&#8203;::UTF8Toggle\->new\("N" x 8\);

format OUT26a =
'^\<\<\<\<\<\<\<\<'~~
$u26a
\.

is\_format\_utf8\(\\\*OUT26a\,
&nbsp;               "'NNNNNNNN '\\n"\);


my $u26b = OS&#8203;::UTF8Toggle\->new\("N" x 8\, 1\);

format OUT26b =
'^\<\<\<\<\<\<\<\<'~~
$u26b
\.

is\_format\_utf8\(\\\*OUT26b\,
&nbsp;               "'NNNNNNNN '\\n"\);

my $u26c = OS&#8203;::UTF8Toggle\->new\("\\x\{FF\}" x 8\);

format OUT26c =
'^\<\<\<\<\<\<\<\<'~~
$u26c
\.

is\_format\_utf8\(\\\*OUT26c\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\} '\\n"\);

my $u26d = OS&#8203;::UTF8Toggle\->new\("\\x\{FF\}" x 8\, 1\);

format OUT26d =
'^\<\<\<\<\<\<\<\<'~~
$u26d
\.

is\_format\_utf8\(\\\*OUT26d\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\} '\\n"\);


my $u26f = OS&#8203;::UTF8Toggle\->new\("N" x 8\);

format OUT26f =
'^\<'~~
$u26f
\.

is\_format\_utf8\(\\\*OUT26f\,
&nbsp;               "'NN'\\n"x4\);


my $u26g = OS&#8203;::UTF8Toggle\->new\("N" x 8\, 1\);

format OUT26g =
'^\<'~~
$u26g
\.

is\_format\_utf8\(\\\*OUT26g\,
&nbsp;               "'NN'\\n"x4\);

my $u26h = OS&#8203;::UTF8Toggle\->new\("\\x\{FF\}" x 8\);

format OUT26h =
'^\<'~~
$u26h
\.

is\_format\_utf8\(\\\*OUT26h\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}'\\n"x4\);

my $u26i = OS&#8203;::UTF8Toggle\->new\("\\x\{FF\}" x 8\, 1\);

format OUT26i =
'^\<'~~
$u26i
\.

is\_format\_utf8\(\\\*OUT26i\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}'\\n"x4\);

**Comment by p5pRT** [comment]
### From @nwc10

\#\!\./perl

BEGIN \{
&nbsp;    chdir 't' if \-d 't';
&nbsp;    @&#8203;INC = '\.\./lib';
&nbsp;    require '\./test\.pl';
\}

use strict;

\# read in a file
sub cat\_utf8 \{
&nbsp;    my $file = shift;
&nbsp;    local $/;
&nbsp;    open my $fh\, '\<'\, $file or die "can't open '$file'&#8203;: $\!";
&nbsp;    binmode $fh\, '&#8203;:utf8';
&nbsp;    my $data = \<$fh> // die "Can't read from '$file'&#8203;: $\!";
&nbsp;    close $fh or die "error closing '$file'&#8203;: $\!";
&nbsp;    $data;
\}

sub is\_format\_utf8 \{
&nbsp;    my \($glob\, $want\, $desc\) = @&#8203;\_;
&nbsp;    local $&#8203;::Level = $&#8203;::Level \+ 1;
&nbsp;    my $file = 'Op\_write\.tmp';
&nbsp;    open $glob\, '>&#8203;:utf8'\, $file or die "Can't create '$file'&#8203;: $\!";
&nbsp;    write $glob;
&nbsp;    close $glob or die "Could not close '$file'&#8203;: $\!";
&nbsp;    is\(cat\_utf8\($file\)\, $want\, $desc\);
\}

plan\(10\);
watchdog\(5\);

\{
&nbsp;    package OS;
&nbsp;    use overload "" => sub \{ $\{$\_\[0\]\}; \};

&nbsp;    sub new \{
&nbsp;        my \($class\, $value\) = @&#8203;\_;
&nbsp;        bless \\$value\, $class;
&nbsp;    \}
\}

my $u25a = OS\->new\("N" x 8\);

format OUT25a =
'^\<\<\<\<\<\<\<\<'~~
$u25a
\.

is\_format\_utf8\(\\\*OUT25a\,
&nbsp;               "'NNNNNNNN '\\n"\);


my $temp = "N" x 8;
utf8&#8203;::upgrade\($temp\);
my $u25b = OS\->new\($temp\);

format OUT25b =
'^\<\<\<\<\<\<\<\<'~~
$u25b
\.

is\_format\_utf8\(\\\*OUT25b\,
&nbsp;               "'NNNNNNNN '\\n"\);

my $u25c = OS\->new\("\\x\{FF\}" x 8\);

format OUT25c =
'^\<\<\<\<\<\<\<\<'~~
$u25c
\.

is\_format\_utf8\(\\\*OUT25c\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\} '\\n"\);

$temp = "\\x\{FF\}" x 8;
utf8&#8203;::upgrade\($temp\);
my $u25d = OS\->new\($temp\);

format OUT25d =
'^\<\<\<\<\<\<\<\<'~~
$u25d
\.

is\_format\_utf8\(\\\*OUT25d\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\}\\x\{FF\} '\\n"\);

my $u25e = OS\->new\("\\x\{100\}" x 8\);

format OUT25e =
'^\<\<\<\<\<\<\<\<'~~
$u25e
\.

is\_format\_utf8\(\\\*OUT25e\,
&nbsp;               "'\\x\{100\}\\x\{100\}\\x\{100\}\\x\{100\}\\x\{100\}\\x\{100\}\\x\{100\}\\x\{100\} '\\n"\);


my $u25f = OS\->new\("N" x 8\);

format OUT25f =
'^\<'~~
$u25f
\.

is\_format\_utf8\(\\\*OUT25f\,
&nbsp;               "'NN'\\n"x4\);


$temp = "N" x 8;
utf8&#8203;::upgrade\($temp\);
my $u25g = OS\->new\($temp\);

format OUT25g =
'^\<'~~
$u25g
\.

is\_format\_utf8\(\\\*OUT25g\,
&nbsp;               "'NN'\\n"x4\);

my $u25h = OS\->new\("\\x\{FF\}" x 8\);

format OUT25h =
'^\<'~~
$u25h
\.

is\_format\_utf8\(\\\*OUT25h\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}'\\n"x4\);

$temp = "\\x\{FF\}" x 8;
utf8&#8203;::upgrade\($temp\);
my $u25i = OS\->new\($temp\);

format OUT25i =
'^\<'~~
$u25i
\.

is\_format\_utf8\(\\\*OUT25i\,
&nbsp;               "'\\x\{FF\}\\x\{FF\}'\\n"x4\);

my $u25j = OS\->new\("\\x\{100\}" x 8\);

format OUT25j =
'^\<'~~
$u25j
\.

is\_format\_utf8\(\\\*OUT25j\,
&nbsp;               "'\\x\{100\}\\x\{100\}'\\n"x4\);

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Sep 16\, 2013 at 05&#8203;:40&#8203;:17AM \-0700\, Nicholas Clark wrote&#8203;:
> formats loop forever chewing RAM for non SVPVs\. This isn't a recent problem\,
> but a couple of \(relatively\) recent commits have exacerbated it\.

\[snip\]
> There is a non\-zero possibility that existing user code is at risk due to
> this\, if upgrading from pre\-5\.12 to 5\.12\, because it will no longer croak\(\)
> for some problem cases\. I can't think of a case where existing code that
> worked cleanly stops working\, and I think that all upgraded code would be
> spotted with sufficient \(probably any\) testing\. Hence I'm wondering if the
> risk of subverting this bug is theoretical\, and hence we should treat this
> the same way as we handled the bug fixed by&#8203;:
> 
> commit 35c6393c7abe46176380cd5f437568a98c8f4995
> Author&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
> Date&#8203;:   Sun Jul 13 21&#8203;:37&#8203;:21 2008 \+0000
> 
>     Make format items @&#8203;\* and ^\* work with references \(safely\)\. Note no\-one
>     said anything about sanely\.
>     
>     p4raw\-id&#8203;: //depot/perl@&#8203;34140
> 
> 
> \(ie we just fixed it\)

Agreed\. I don't think its a security issue\. I have a working patch\,
which incorporates all Nicholas' test code\.

> Given that I was thinking that the solution is Perl\_sv\_setpvn\(\)\, I was
> wondering what happens if the SV is not the same UTF8\-ness each time\, which
> turned out to produce interesting results\. I think that it might be fixed
> \(or at least partly fixed\) when the sv\_chop\(\) problem is addressed\. However\,
> I'll report it as a separate bug\, because it seems to have fewer untoward
> side effects\. \(The combination of the UTF\-8 toggle and an \*overloaded\*
> string is a RAM\-chewer\, so I'll attach its tests to this report\)

My patch fixes this too\.

Unless anyone objects\, I'll push this fix publicly in a couple of days\.


Here's the complete commit&#8203;:

commit adccfec71f2d1fcea29f854152774368529435e1
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Thu Nov 7 12&#8203;:17&#8203;:26 2013 \+0000
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Thu Nov 7 15&#8203;:20&#8203;:45 2013 \+0000

&nbsp;    fix chop formats with non PV vars
&nbsp;    
&nbsp;    \[perl \#119847\]\,  \[perl \#119849\]\, \[perl \#119851\]
&nbsp;    
&nbsp;    Strange vars like ties\, overloads\, or stringified refs \(and in recent
&nbsp;    perls\, pure NOK vars\) would generally do the wrong thing in formats
&nbsp;    when the var is treated as a string and repeatedly chopped\, as in
&nbsp;    ^\<\<\<~~ and similar\. This would manifest itself in infinite loops\, utf8
&nbsp;    errors etc\. A recent change that stopped a stringified NOK getting
&nbsp;    converted into a POK made the same badness happen for plain NVs too\.
&nbsp;    
&nbsp;    This commit contains two main fixes\. First\, the chopping was done
&nbsp;    using sv\_chop\(\)\, which only worked on POK strings\. If its \!POK\, we now do
&nbsp;    sv\_setpvn\(\) instead\, which is less efficient\, but will ensure the right
&nbsp;    thing is always done\.
&nbsp;    
&nbsp;    Secondly\, we make sure that the sv is accessed only once per cycle\,
&nbsp;    doing s = SvPV\(sv\, len\) or similar\. After that\, all access is done only
&nbsp;    via s and len\. One place was using SvPVX\(sv\)\, and several places
&nbsp;    were using the sv for utf8\<\->byte length conversions\, such as
&nbsp;    sv\_pos\_b2u\(\)\.
&nbsp;    
&nbsp;    It turns out that all the complex utf8 handling could be enormously
&nbsp;    simplified\. Since the code that needed to do utf8/byte length conversions
&nbsp;    already scanned the string looking for suitable split points \(such as
&nbsp;    spaces or \\n or \\r\)\, it was easiest to include any utf8 processing in the
&nbsp;    same loop \- i\.e\. incrementing s by UTF8SKIP\(s\) each time\, but incrementing
&nbsp;    the character count by 1\.
&nbsp;    
&nbsp;    The original diagnosis and reporting of this issue was done by Nicholas
&nbsp;    Clark\, who also supplied most of the tests\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    pp\_ctl\.c
&nbsp;    M    t/op/write\.t

Differences \.\.\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/pp_ctl.c b/pp_ctl.c
index 95572aa..f9d6821 100644
--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -466,7 +466,8 @@ PP(pp_formline)
     I32 arg;
     SV *sv = NULL; /* current item */
     const char *item = NULL;/* string value of current item */
-    I32 itemsize  = 0;	    /* length of current item, possibly truncated */
+    I32 itemsize  = 0;	    /* length (chars) of item, possibly truncated */
+    I32 itembytes = 0;	    /* as itemsize, but length in bytes */
     I32 fieldsize = 0;	    /* width of current field */
     I32 lines = 0;	    /* number of lines that have been output */
     bool chopspace = (strchr(PL_chopset, ' ') != NULL); /* does $: have space */
@@ -474,7 +475,7 @@ PP(pp_formline)
     STRLEN linemark = 0;    /* pos of start of line in output */
     NV value;
     bool gotsome = FALSE;   /* seen at least one non-blank item on this line */
-    STRLEN len;
+    STRLEN len;             /* length of current sv */
     STRLEN linemax;	    /* estimate of output size in bytes */
     bool item_is_utf8 = FALSE;
     bool targ_is_utf8 = FALSE;
@@ -569,133 +570,85 @@ PP(pp_formline)
 
 	case FF_CHECKNL: /* find max len of item (up to \n) that fits field */
 	    {
-		const char *send;
 		const char *s = item = SvPV_const(sv, len);
-		itemsize = len;
-		if (DO_UTF8(sv)) {
-		    itemsize = sv_len_utf8(sv);
-		    if (itemsize != (I32)len) {
-			I32 itembytes;
-			if (itemsize > fieldsize) {
-			    itemsize = fieldsize;
-			    itembytes = itemsize;
-			    sv_pos_u2b(sv, &itembytes, 0);
-			}
-			else
-			    itembytes = len;
-			send = chophere = s + itembytes;
-			while (s < send) {
-			    if (! isCNTRL(*s))
-				gotsome = TRUE;
-			    else if (*s == '\n')
-				break;
-			    s++;
-			}
-			item_is_utf8 = TRUE;
-			itemsize = s - item;
-			sv_pos_b2u(sv, &itemsize);
-			break;
-		    }
-		}
-		item_is_utf8 = FALSE;
-		if (itemsize > fieldsize)
-		    itemsize = fieldsize;
-		send = chophere = s + itemsize;
-		while (s < send) {
-		    if (! isCNTRL(*s))
-			gotsome = TRUE;
-		    else if (*s == '\n')
-			break;
-		    s++;
-		}
-		itemsize = s - item;
+		const char *send = s + len;
+
+                itemsize = 0;
+		item_is_utf8 = DO_UTF8(sv);
+                while (s < send) {
+                    if (!isCNTRL(*s))
+                        gotsome = TRUE;
+                    else if (*s == '\n')
+                        break;
+
+                    if (item_is_utf8)
+                        s += UTF8SKIP(s);
+                    else
+                        s++;
+                    itemsize++;
+                    if (itemsize == fieldsize)
+                        break;
+                }
+                itembytes = s - item;
 		break;
 	    }
 
 	case FF_CHECKCHOP: /* like CHECKNL, but up to highest split point */
 	    {
 		const char *s = item = SvPV_const(sv, len);
-		itemsize = len;
-		if (DO_UTF8(sv)) {
-		    itemsize = sv_len_utf8(sv);
-		    if (itemsize != (I32)len) {
-			I32 itembytes;
-			if (itemsize <= fieldsize) {
-			    const char *send = chophere = s + itemsize;
-			    while (s < send) {
-				if (*s == '\r') {
-				    itemsize = s - item;
-				    chophere = s;
-				    break;
-				}
-				if (! isCNTRL(*s))
-				    gotsome = TRUE;
-                                s++;
-			    }
-			}
-			else {
-			    const char *send;
-			    itemsize = fieldsize;
-			    itembytes = itemsize;
-			    sv_pos_u2b(sv, &itembytes, 0);
-			    send = chophere = s + itembytes;
-			    while (s < send || (s == send && isSPACE(*s))) {
-				if (isSPACE(*s)) {
-				    if (chopspace)
-					chophere = s;
-				    if (*s == '\r')
-					break;
-				}
-				else {
-				    if (! isCNTRL(*s))
-					gotsome = TRUE;
-				    if (strchr(PL_chopset, *s))
-					chophere = s + 1;
-				}
-				s++;
-			    }
-			    itemsize = chophere - item;
-			    sv_pos_b2u(sv, &itemsize);
-			}
-			item_is_utf8 = TRUE;
-			break;
-		    }
-		}
-		item_is_utf8 = FALSE;
-		if (itemsize <= fieldsize) {
-		    const char *const send = chophere = s + itemsize;
-		    while (s < send) {
-			if (*s == '\r') {
-			    itemsize = s - item;
-			    chophere = s;
-			    break;
-			}
-			if (! isCNTRL(*s))
-			    gotsome = TRUE;
+		const char *send = s + len;
+                I32 size = 0;
+
+                chophere = NULL;
+		item_is_utf8 = DO_UTF8(sv);
+                while (s < send) {
+                    /* look for a legal split position */
+                    if (isSPACE(*s)) {
+                        if (*s == '\r') {
+                            chophere = s;
+                            itemsize = size;
+                            break;
+                        }
+                        if (chopspace) {
+                            /* provisional split point */
+                            chophere = s;
+                            itemsize = size;
+                        }
+                        /* we delay testing fieldsize until after we've
+                         * processed the possible split char directly
+                         * following the last field char; so if fieldsize=3
+                         * and item="a b cdef", we consume "a b", not "a".
+                         * Ditto further down.
+                         */
+                        if (size == fieldsize)
+                            break;
+                    }
+                    else {
+                        if (strchr(PL_chopset, *s)) {
+                            /* provisional split point */
+                            /* for a non-space split char, we include
+                             * the split char; hence the '+1' */
+                            chophere = s + 1;
+                            itemsize = size;
+                        }
+                        if (size == fieldsize)
+                            break;
+                        if (!isCNTRL(*s))
+                            gotsome = TRUE;
+                    }
+
+                    if (item_is_utf8)
+                        s += UTF8SKIP(s);
+                    else
                         s++;
-		    }
-		}
-		else {
-		    const char *send;
-		    itemsize = fieldsize;
-		    send = chophere = s + itemsize;
-		    while (s < send || (s == send && isSPACE(*s))) {
-			if (isSPACE(*s)) {
-			    if (chopspace)
-				chophere = s;
-			    if (*s == '\r')
-				break;
-			}
-			else {
-			    if (! isCNTRL(*s))
-				gotsome = TRUE;
-			    if (strchr(PL_chopset, *s))
-				chophere = s + 1;
-			}
-			s++;
-		    }
-		    itemsize = chophere - item;
-		}
+                    size++;
+                }
+                if (!chophere || s == send) {
+                    chophere = s;
+                    itemsize = size;
+                }
+                itembytes = chophere - item;
+
 		break;
 	    }
 
@@ -719,16 +672,9 @@ PP(pp_formline)
 	    break;
 
 	case FF_ITEM: /* append a text item, while blanking ctrl chars */
-	    to_copy = itemsize;
+	    to_copy = itembytes;
 	    source = (U8 *)item;
 	    trans = 1;
-	    if (item_is_utf8) {
-		/* convert to_copy from chars to bytes */
-		U8 *s = source;
-		while (to_copy--)
-		   s += UTF8SKIP(s);
-		to_copy = s - source;
-	    }
 	    goto append;
 
 	case FF_CHOP: /* (for ^*) chop the current item */
@@ -738,7 +684,12 @@ PP(pp_formline)
 		    while (isSPACE(*s))
 			s++;
 		}
-		sv_chop(sv,s);
+                if (SvPOKp(sv))
+                    sv_chop(sv,s);
+                else
+                    /* tied, overloaded or similar strangeness.
+                     * Do it the hard way */
+                    sv_setpvn(sv, s, len - (s-item));
 		SvSETMAGIC(sv);
 		break;
 	    }
@@ -763,7 +714,7 @@ PP(pp_formline)
 		while (s < send) {
 		    if (*s++ == '\n') {
 			if (oneline) {
-			    to_copy = s - SvPVX_const(sv) - 1;
+			    to_copy = s - item - 1;
 			    chophere = s;
 			    break;
 			} else {
diff --git a/t/op/write.t b/t/op/write.t
index 7dcf1e0..0bad2e4 100644
--- a/t/op/write.t
+++ b/t/op/write.t
@@ -7,6 +7,7 @@ BEGIN {
 }
 
 use strict;	# Amazed that this hackery can be made strict ...
+use Tie::Scalar;
 
 # read in a file
 sub cat {
@@ -18,6 +19,42 @@ sub cat {
     $data;
 }
 
+# read in a utf-8 file
+#
+sub cat_utf8 {
+    my $file = shift;
+    local $/;
+    open my $fh, '<', $file or die "can't open '$file': $!";
+    binmode $fh, ':utf8';
+    my $data = <$fh> // die "Can't read from '$file': $!";
+    close $fh or die "error closing '$file': $!";
+    $data;
+}
+
+# write a format to a utf8 file, then read it back in and compare
+
+sub is_format_utf8 {
+    my ($glob, $want, $desc) = @_;
+    local $::Level = $::Level + 1;
+    my $file = 'Op_write.tmp';
+    open $glob, '>:utf8', $file or die "Can't create '$file': $!";
+    write $glob;
+    close $glob or die "Could not close '$file': $!";
+    is(cat_utf8($file), $want, $desc);
+}
+
+sub like_format_utf8 {
+    my ($glob, $want, $desc) = @_;
+    local $::Level = $::Level + 1;
+    my $file = 'Op_write.tmp';
+    open $glob, '>:utf8', $file or die "Can't create '$file': $!";
+    write $glob;
+    close $glob or die "Could not close '$file': $!";
+    like(cat_utf8($file), $want, $desc);
+}
+
+
+
 #-- testing numeric fields in all variants (WL)
 
 sub swrite {
@@ -61,7 +98,7 @@ for my $tref ( @NumTests ){
 my $bas_tests = 21;
 
 # number of tests in section 3
-my $bug_tests = 8 + 3 * 3 * 5 * 2 * 3 + 2 + 66 + 4 + 2 + 3 + 96 + 11;
+my $bug_tests = 66 + 3 * 3 * 5 * 2 * 3 + 2 + 66 + 4 + 2 + 3 + 96 + 11;
 
 # number of tests in section 4
 my $hmb_tests = 35;
@@ -351,7 +388,6 @@ $el
 {
     my $test = curr_test();
     # Bug report and testcase by Alexey Tourbin
-    use Tie::Scalar;
     my $v;
     tie $v, 'Tie::StdScalar';
     $v = $test;
@@ -560,6 +596,770 @@ $_
     close OUT21 or die "Could not close: $!";
 })[0];
 
+
+
+# [perl #119847],  [perl #119849], [perl #119851]
+# Non-real vars like tied, overloaded and refs could, when stringified,
+# fail to be processed properly, causing infinite loops on ~~, utf8
+# warnings etc, ad nauseum.
+
+
+my $u22a = "N" x 8;
+
+format OUT22a =
+'^<<<<<<<<'~~
+$u22a
+.
+
+is_format_utf8(\*OUT22a,
+               "'NNNNNNNN '\n");
+
+
+my $u22b = "N" x 8;
+utf8::upgrade($u22b);
+
+format OUT22b =
+'^<<<<<<<<'~~
+$u22b
+.
+
+is_format_utf8(\*OUT22b,
+               "'NNNNNNNN '\n");
+
+my $u22c = "\x{FF}" x 8;
+
+format OUT22c =
+'^<<<<<<<<'~~
+$u22c
+.
+
+is_format_utf8(\*OUT22c,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+my $u22d = "\x{FF}" x 8;
+utf8::upgrade($u22d);
+
+format OUT22d =
+'^<<<<<<<<'~~
+$u22d
+.
+
+is_format_utf8(\*OUT22d,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+my $u22e = "\x{100}" x 8;
+
+format OUT22e =
+'^<<<<<<<<'~~
+$u22e
+.
+
+is_format_utf8(\*OUT22e,
+               "'\x{100}\x{100}\x{100}\x{100}\x{100}\x{100}\x{100}\x{100} '\n");
+
+
+my $u22f = "N" x 8;
+
+format OUT22f =
+'^<'~~
+$u22f
+.
+
+is_format_utf8(\*OUT22f,
+               "'NN'\n"x4);
+
+
+my $u22g = "N" x 8;
+utf8::upgrade($u22g);
+
+format OUT22g =
+'^<'~~
+$u22g
+.
+
+is_format_utf8(\*OUT22g,
+               "'NN'\n"x4);
+
+my $u22h = "\x{FF}" x 8;
+
+format OUT22h =
+'^<'~~
+$u22h
+.
+
+is_format_utf8(\*OUT22h,
+               "'\x{FF}\x{FF}'\n"x4);
+
+my $u22i = "\x{FF}" x 8;
+utf8::upgrade($u22i);
+
+format OUT22i =
+'^<'~~
+$u22i
+.
+
+is_format_utf8(\*OUT22i,
+               "'\x{FF}\x{FF}'\n"x4);
+
+my $u22j = "\x{100}" x 8;
+
+format OUT22j =
+'^<'~~
+$u22j
+.
+
+is_format_utf8(\*OUT22j,
+               "'\x{100}\x{100}'\n"x4);
+
+
+tie my $u23a, 'Tie::StdScalar';
+$u23a = "N" x 8;
+
+format OUT23a =
+'^<<<<<<<<'~~
+$u23a
+.
+
+is_format_utf8(\*OUT23a,
+               "'NNNNNNNN '\n");
+
+
+tie my $u23b, 'Tie::StdScalar';
+$u23b = "N" x 8;
+utf8::upgrade($u23b);
+
+format OUT23b =
+'^<<<<<<<<'~~
+$u23b
+.
+
+is_format_utf8(\*OUT23b,
+               "'NNNNNNNN '\n");
+
+tie my $u23c, 'Tie::StdScalar';
+$u23c = "\x{FF}" x 8;
+
+format OUT23c =
+'^<<<<<<<<'~~
+$u23c
+.
+
+is_format_utf8(\*OUT23c,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+tie my $u23d, 'Tie::StdScalar';
+my $temp = "\x{FF}" x 8;
+utf8::upgrade($temp);
+$u23d = $temp;
+
+format OUT23d =
+'^<<<<<<<<'~~
+$u23d
+.
+
+is_format_utf8(\*OUT23d,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+tie my $u23e, 'Tie::StdScalar';
+$u23e = "\x{100}" x 8;
+
+format OUT23e =
+'^<<<<<<<<'~~
+$u23e
+.
+
+is_format_utf8(\*OUT23e,
+               "'\x{100}\x{100}\x{100}\x{100}\x{100}\x{100}\x{100}\x{100} '\n");
+
+tie my $u23f, 'Tie::StdScalar';
+$u23f = "N" x 8;
+
+format OUT23f =
+'^<'~~
+$u23f
+.
+
+is_format_utf8(\*OUT23f,
+               "'NN'\n"x4);
+
+
+tie my $u23g, 'Tie::StdScalar';
+my $temp = "N" x 8;
+utf8::upgrade($temp);
+$u23g = $temp;
+
+format OUT23g =
+'^<'~~
+$u23g
+.
+
+is_format_utf8(\*OUT23g,
+               "'NN'\n"x4);
+
+tie my $u23h, 'Tie::StdScalar';
+$u23h = "\x{FF}" x 8;
+
+format OUT23h =
+'^<'~~
+$u23h
+.
+
+is_format_utf8(\*OUT23h,
+               "'\x{FF}\x{FF}'\n"x4);
+
+$temp = "\x{FF}" x 8;
+utf8::upgrade($temp);
+tie my $u23i, 'Tie::StdScalar';
+$u23i = $temp;
+
+format OUT23i =
+'^<'~~
+$u23i
+.
+
+is_format_utf8(\*OUT23i,
+               "'\x{FF}\x{FF}'\n"x4);
+
+tie my $u23j, 'Tie::StdScalar';
+$u23j = "\x{100}" x 8;
+
+format OUT23j =
+'^<'~~
+$u23j
+.
+
+is_format_utf8(\*OUT23j,
+               "'\x{100}\x{100}'\n"x4);
+
+{
+    package UTF8Toggle;
+
+    sub TIESCALAR {
+        my $class = shift;
+        my $value = shift;
+        my $state = shift||0;
+        return bless [$value, $state], $class;
+    }
+
+    sub FETCH {
+        my $self = shift;
+        $self->[1] = ! $self->[1];
+        if ($self->[1]) {
+           utf8::downgrade($self->[0]);
+        } else {
+           utf8::upgrade($self->[0]);
+        }
+        $self->[0];
+    }
+
+   sub STORE {
+       my $self = shift;
+       $self->[0] = shift;
+    }
+}
+
+tie my $u24a, 'UTF8Toggle';
+$u24a = "N" x 8;
+
+format OUT24a =
+'^<<<<<<<<'~~
+$u24a
+.
+
+is_format_utf8(\*OUT24a,
+               "'NNNNNNNN '\n");
+
+
+tie my $u24b, 'UTF8Toggle';
+$u24b = "N" x 8;
+utf8::upgrade($u24b);
+
+format OUT24b =
+'^<<<<<<<<'~~
+$u24b
+.
+
+is_format_utf8(\*OUT24b,
+               "'NNNNNNNN '\n");
+
+tie my $u24c, 'UTF8Toggle';
+$u24c = "\x{FF}" x 8;
+
+format OUT24c =
+'^<<<<<<<<'~~
+$u24c
+.
+
+is_format_utf8(\*OUT24c,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+tie my $u24d, 'UTF8Toggle', 1;
+$u24d = "\x{FF}" x 8;
+
+format OUT24d =
+'^<<<<<<<<'~~
+$u24d
+.
+
+is_format_utf8(\*OUT24d,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+
+
+tie my $u24f, 'UTF8Toggle';
+$u24f = "N" x 8;
+
+format OUT24f =
+'^<'~~
+$u24f
+.
+
+is_format_utf8(\*OUT24f,
+               "'NN'\n"x4);
+
+
+tie my $u24g, 'UTF8Toggle';
+my $temp = "N" x 8;
+utf8::upgrade($temp);
+$u24g = $temp;
+
+format OUT24g =
+'^<'~~
+$u24g
+.
+
+is_format_utf8(\*OUT24g,
+               "'NN'\n"x4);
+
+tie my $u24h, 'UTF8Toggle';
+$u24h = "\x{FF}" x 8;
+
+format OUT24h =
+'^<'~~
+$u24h
+.
+
+is_format_utf8(\*OUT24h,
+               "'\x{FF}\x{FF}'\n"x4);
+
+tie my $u24i, 'UTF8Toggle', 1;
+$u24i = "\x{FF}" x 8;
+
+format OUT24i =
+'^<'~~
+$u24i
+.
+
+is_format_utf8(\*OUT24i,
+               "'\x{FF}\x{FF}'\n"x4);
+
+{
+    package OS;
+    use overload '""' => sub { ${$_[0]}; };
+
+    sub new {
+        my ($class, $value) = @_;
+        bless \$value, $class;
+    }
+}
+
+my $u25a = OS->new("N" x 8);
+
+format OUT25a =
+'^<<<<<<<<'~~
+$u25a
+.
+
+is_format_utf8(\*OUT25a,
+               "'NNNNNNNN '\n");
+
+
+my $temp = "N" x 8;
+utf8::upgrade($temp);
+my $u25b = OS->new($temp);
+
+format OUT25b =
+'^<<<<<<<<'~~
+$u25b
+.
+
+is_format_utf8(\*OUT25b,
+               "'NNNNNNNN '\n");
+
+my $u25c = OS->new("\x{FF}" x 8);
+
+format OUT25c =
+'^<<<<<<<<'~~
+$u25c
+.
+
+is_format_utf8(\*OUT25c,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+$temp = "\x{FF}" x 8;
+utf8::upgrade($temp);
+my $u25d = OS->new($temp);
+
+format OUT25d =
+'^<<<<<<<<'~~
+$u25d
+.
+
+is_format_utf8(\*OUT25d,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+my $u25e = OS->new("\x{100}" x 8);
+
+format OUT25e =
+'^<<<<<<<<'~~
+$u25e
+.
+
+is_format_utf8(\*OUT25e,
+               "'\x{100}\x{100}\x{100}\x{100}\x{100}\x{100}\x{100}\x{100} '\n");
+
+
+my $u25f = OS->new("N" x 8);
+
+format OUT25f =
+'^<'~~
+$u25f
+.
+
+is_format_utf8(\*OUT25f,
+               "'NN'\n"x4);
+
+
+$temp = "N" x 8;
+utf8::upgrade($temp);
+my $u25g = OS->new($temp);
+
+format OUT25g =
+'^<'~~
+$u25g
+.
+
+is_format_utf8(\*OUT25g,
+               "'NN'\n"x4);
+
+my $u25h = OS->new("\x{FF}" x 8);
+
+format OUT25h =
+'^<'~~
+$u25h
+.
+
+is_format_utf8(\*OUT25h,
+               "'\x{FF}\x{FF}'\n"x4);
+
+$temp = "\x{FF}" x 8;
+utf8::upgrade($temp);
+my $u25i = OS->new($temp);
+
+format OUT25i =
+'^<'~~
+$u25i
+.
+
+is_format_utf8(\*OUT25i,
+               "'\x{FF}\x{FF}'\n"x4);
+
+my $u25j = OS->new("\x{100}" x 8);
+
+format OUT25j =
+'^<'~~
+$u25j
+.
+
+is_format_utf8(\*OUT25j,
+               "'\x{100}\x{100}'\n"x4);
+
+{
+    package OS::UTF8Toggle;
+    use overload '""' => sub {
+        my $self = shift;
+        $self->[1] = ! $self->[1];
+        if ($self->[1]) {
+            utf8::downgrade($self->[0]);
+        } else {
+            utf8::upgrade($self->[0]);
+        }
+        $self->[0];
+    };
+
+    sub new {
+        my ($class, $value, $state) = @_;
+        bless [$value, $state], $class;
+    }
+}
+
+
+my $u26a = OS::UTF8Toggle->new("N" x 8);
+
+format OUT26a =
+'^<<<<<<<<'~~
+$u26a
+.
+
+is_format_utf8(\*OUT26a,
+               "'NNNNNNNN '\n");
+
+
+my $u26b = OS::UTF8Toggle->new("N" x 8, 1);
+
+format OUT26b =
+'^<<<<<<<<'~~
+$u26b
+.
+
+is_format_utf8(\*OUT26b,
+               "'NNNNNNNN '\n");
+
+my $u26c = OS::UTF8Toggle->new("\x{FF}" x 8);
+
+format OUT26c =
+'^<<<<<<<<'~~
+$u26c
+.
+
+is_format_utf8(\*OUT26c,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+my $u26d = OS::UTF8Toggle->new("\x{FF}" x 8, 1);
+
+format OUT26d =
+'^<<<<<<<<'~~
+$u26d
+.
+
+is_format_utf8(\*OUT26d,
+               "'\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF}\x{FF} '\n");
+
+
+my $u26f = OS::UTF8Toggle->new("N" x 8);
+
+format OUT26f =
+'^<'~~
+$u26f
+.
+
+is_format_utf8(\*OUT26f,
+               "'NN'\n"x4);
+
+
+my $u26g = OS::UTF8Toggle->new("N" x 8, 1);
+
+format OUT26g =
+'^<'~~
+$u26g
+.
+
+is_format_utf8(\*OUT26g,
+               "'NN'\n"x4);
+
+my $u26h = OS::UTF8Toggle->new("\x{FF}" x 8);
+
+format OUT26h =
+'^<'~~
+$u26h
+.
+
+is_format_utf8(\*OUT26h,
+               "'\x{FF}\x{FF}'\n"x4);
+
+my $u26i = OS::UTF8Toggle->new("\x{FF}" x 8, 1);
+
+format OUT26i =
+'^<'~~
+$u26i
+.
+
+is_format_utf8(\*OUT26i,
+               "'\x{FF}\x{FF}'\n"x4);
+
+
+
+{
+    my $zero = $$ - $$;
+
+    package Number;
+
+    sub TIESCALAR {
+        my $class = shift;
+        my $value = shift;
+        return bless \$value, $class;
+    }
+
+    # The return value should always be SvNOK() only:
+    sub FETCH {
+        my $self = shift;
+        # avoid "" getting converted to "0" and thus
+        # causing an infinite loop
+        return "" unless length ($$self);
+        return $$self - 0.5 + $zero + 0.5;
+    }
+
+   sub STORE {
+       my $self = shift;
+       $$self = shift;
+    }
+
+   package ONumber;
+
+   use overload '""' => sub {
+        my $self = shift;
+        return $$self - 0.5 + $zero + 0.5;
+    };
+
+    sub new {
+       my $class = shift;
+       my $value = shift;
+       return bless \$value, $class;
+   }
+}
+
+my $v27a = 1/256;
+
+format OUT27a =
+'^<<<<<<<<<'~~
+$v27a
+.
+
+is_format_utf8(\*OUT27a,
+               "'0.00390625'\n");
+
+my $v27b = 1/256;
+
+format OUT27b =
+'^<'~~
+$v27b
+.
+
+is_format_utf8(\*OUT27b,
+               "'0.'\n'00'\n'39'\n'06'\n'25'\n");
+
+tie my $v27c, 'Number', 1/256;
+
+format OUT27c =
+'^<<<<<<<<<'~~
+$v27c
+.
+
+is_format_utf8(\*OUT27c,
+               "'0.00390625'\n");
+
+my $v27d = 1/256;
+
+format OUT27d =
+'^<'~~
+$v27d
+.
+
+is_format_utf8(\*OUT27d,
+               "'0.'\n'00'\n'39'\n'06'\n'25'\n");
+
+my $v27e = ONumber->new(1/256);
+
+format OUT27e =
+'^<<<<<<<<<'~~
+$v27e
+.
+
+is_format_utf8(\*OUT27e,
+               "'0.00390625'\n");
+
+my $v27f = ONumber->new(1/256);
+
+format OUT27f =
+'^<'~~
+$v27f
+.
+
+is_format_utf8(\*OUT27f,
+               "'0.'\n'00'\n'39'\n'06'\n'25'\n");
+
+{
+    package Ref;
+    use overload '""' => sub {
+	return ${$_[0]};
+    };
+
+    sub new {
+       my $class = shift;
+       my $value = shift;
+       return bless \$value, $class;
+   }
+}
+
+my $v28a = {};
+
+format OUT28a =
+'^<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<'~~
+$v28a
+.
+
+
+# 'HASH(0x1716b60)     '
+my $qr_hash   = qr/^'HASH\(0x[0-9a-f]+\)\s+'\n$/;
+
+# 'HASH'
+# '(0x1'
+# '716b'
+# 'c0) '
+my $qr_hash_m = qr/^'HASH'\n('[0-9a-fx() ]{4}'\n)+$/;
+
+like_format_utf8(\*OUT28a, $qr_hash);
+
+my $v28b = {};
+
+format OUT28b =
+'^<<<'~~
+$v28b
+.
+
+like_format_utf8(\*OUT28b, $qr_hash_m);
+
+
+tie my $v28c, 'Tie::StdScalar';
+$v28c = {};
+
+format OUT28c =
+'^<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<'~~
+$v28c
+.
+
+like_format_utf8(\*OUT28c, $qr_hash);
+
+tie my $v28d, 'Tie::StdScalar';
+$v28d = {};
+
+format OUT28d =
+'^<<<'~~
+$v28d
+.
+
+like_format_utf8(\*OUT28d, $qr_hash_m);
+
+my $v28e = Ref->new({});
+
+format OUT28e =
+'^<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<'~~
+$v28e
+.
+
+like_format_utf8(\*OUT28e, $qr_hash);
+
+my $v28f = Ref->new({});
+
+format OUT28f =
+'^<<<'~~
+$v28f
+.
+
+like_format_utf8(\*OUT28f, $qr_hash_m);
+
+
+
 {
   package Count;
 
@@ -627,8 +1427,11 @@ $_
   # pp_formline attempts to set SvCUR() on an SVt_RV). I suspect that it will
   # be doing something similarly out of bounds on everything from 5.000
   my $ref = [];
-  is swrite('>^*<', $ref), ">$ref<";
-  is swrite('>@*<', $ref), ">$ref<";
+  my $exp = ">$ref<";
+  is swrite('>^*<', $ref), $exp;
+  $ref = [];
+  my $exp = ">$ref<";
+  is swrite('>@*<', $ref), $exp;
 }
 
 format EMPTY =


-- 
```
</details>

The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rjbs

\* Dave Mitchell \<davem@&#8203;iabyn\.com> \[2013\-11\-07T10&#8203;:43&#8203;:56\]
> Unless anyone objects\, I'll push this fix publicly in a couple of days\.

\+1

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @iabyn

fixed in 9b4bdfd44e0e6d44a447f231c281f967c7ca35c9

**Comment by p5pRT** [comment]
### From @iabyn

fixed in 9b4bdfd44e0e6d44a447f231c281f967c7ca35c9

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Nov 07\, 2013 at 03&#8203;:43&#8203;:56PM \+0000\, Dave Mitchell wrote&#8203;:

> My patch fixes this too\.
> 
> Unless anyone objects\, I'll push this fix publicly in a couple of days\.

Thanks for looking into all these\, and fixing them\.

>     \[perl \#119847\]\,  \[perl \#119849\]\, \[perl \#119851\]

As they are now all resolved\, should these tickets now be moved to the main
RT queue?

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Nov 16\, 2013 at 10&#8203;:46&#8203;:12AM \+0000\, Nicholas Clark wrote&#8203;:
> Thanks for looking into all these\, and fixing them\.
> 
> >     \[perl \#119847\]\,  \[perl \#119849\]\, \[perl \#119851\]
> 
> As they are now all resolved\, should these tickets now be moved to the main
> RT queue?

Yes\.

A few days ago I tried doing that via the RT interface and got an error\.
Been meaning to get round to reporting it\.


\-\- 
"Procrastination grows to fill the available time"
&nbsp;    \-\- Mitchell's corollary to Parkinson's Law

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Nov 16\, 2013 at 01&#8203;:08&#8203;:28PM \+0000\, Dave Mitchell wrote&#8203;:
> On Sat\, Nov 16\, 2013 at 10&#8203;:46&#8203;:12AM \+0000\, Nicholas Clark wrote&#8203;:
> > Thanks for looking into all these\, and fixing them\.
> > 
> > >     \[perl \#119847\]\,  \[perl \#119849\]\, \[perl \#119851\]
> > 
> > As they are now all resolved\, should these tickets now be moved to the main
> > RT queue?
> 
> Yes\.
> 
> A few days ago I tried doing that via the RT interface and got an error\.
> Been meaning to get round to reporting it\.

Reported\, fixed\, and now all moved to perl5 queue

\-\- 
No man treats a motor car as foolishly as he treats another human being\.
When the car will not go\, he does not attribute its annoying behaviour to
sin\, he does not say\, You are a wicked motorcar\, and I shall not give you
any more petrol until you go\. He attempts to find out what is wrong and
set it right\.
&nbsp;    \-\- Bertrand Russell\,
&nbsp;       Has Religion Made Useful Contributions to Civilization?

**Comment by p5pRT** [comment]
### From @nwc10

On Thu Dec 05 04&#8203;:27&#8203;:15 2013\, davem wrote&#8203;:
> On Sat\, Nov 16\, 2013 at 01&#8203;:08&#8203;:28PM \+0000\, Dave Mitchell wrote&#8203;:
> > On Sat\, Nov 16\, 2013 at 10&#8203;:46&#8203;:12AM \+0000\, Nicholas Clark wrote&#8203;:
> > > Thanks for looking into all these\, and fixing them\.
> > >
> > > > \[perl \#119847\]\,  \[perl \#119849\]\, \[perl \#119851\]
> > >
> > > As they are now all resolved\, should these tickets now be moved to
> > > the main
> > > RT queue?
> >
> > Yes\.
> >
> > A few days ago I tried doing that via the RT interface and got an
> > error\.
> > Been meaning to get round to reporting it\.
> 
> Reported\, fixed\, and now all moved to perl5 queue

Confirm that I can see these in the main queue before I logged in\.

Thanks\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu Dec 05 04&#8203;:27&#8203;:15 2013\, davem wrote&#8203;:
> On Sat\, Nov 16\, 2013 at 01&#8203;:08&#8203;:28PM \+0000\, Dave Mitchell wrote&#8203;:
> > On Sat\, Nov 16\, 2013 at 10&#8203;:46&#8203;:12AM \+0000\, Nicholas Clark wrote&#8203;:
> > > Thanks for looking into all these\, and fixing them\.
> > >
> > > > \[perl \#119847\]\,  \[perl \#119849\]\, \[perl \#119851\]
> > >
> > > As they are now all resolved\, should these tickets now be moved to
> > > the main
> > > RT queue?
> >
> > Yes\.
> >
> > A few days ago I tried doing that via the RT interface and got an
> > error\.
> > Been meaning to get round to reporting it\.
> 
> Reported\, fixed\, and now all moved to perl5 queue

Confirm that I can see these in the main queue before I logged in\.

Thanks\.

Nicholas Clark
